### ISSUE TITLE
[Feature]: Compute and log the serving FLOPs

### ISSUE THREAD

**Issue by zhuohan123** [vllm-project/vllm#3490]
### 🚀 The feature, motivation and pitch

vLLM should output the serving FLOPs. This should be helpful for debugging performance and check the GPU utilization.

**Comment by ayusher** [comment]
Happy to take this up!

**Comment by gardberg** [comment]
@ayusher Just out of curiosity, how would you go about doing this? Let's take the simplest case of a single non-sharded model running on a machine. I assume we want to log the actual FLOPs / MACs (multiply accumulates) that the hardware is doing, and not estimate it from the modules of the model e.g. via [profiling ](https://github.com/adityaiitb/PyProf) or just theoretically [counting per-module](https://github.com/Lyken17/pytorch-OpCounter), as this does not take into consideration the kernel implementation of the module (different implementations of attention in e.g. cuda can require different amount of FLOPs, right?). What would be the right way of doing this, and also not introduce too much overhead?

**Comment by ayusher** [comment]
Yeah, after further research this looks like a more involved problem than I initially anticipated. I'm not sure on the best approach or if this is worth considering right now.

**Comment by gardberg** [comment]
Yeah I agree. I know that some nvidia gpu libraries enable you to get flops, but does not seem like a trivial problem (not a good first issue I guess lol). I'd love to help if we find a manageable way though!

**Comment by rakshithvasudev** [comment]
Thank you all! 

Are the flops required to be accurate? Is it okay if the FLOPS numbers are a close estimation? 

As discussed by others in this thread it's not very straightforward to calculate FLOPS without digging into internals. That may also add overheads during inference. If okay with a close estimation, I can try to implement a basic FLOPS calculator.

**Comment by krtkvrm** [comment]
can I take this up?

**Comment by gangula-karthik** [comment]
hi @zhuohan123 is this issue still open to work on ?

**Comment by duhaode520** [comment]
@zhuohan123 @robertgshaw2-redhat @WoosukKwon  Hi, is this issue still open? I'd like to take this if it is still needed.

**Comment by plops655** [comment]
@duhaode520 Do you have any ideas as to how to approach this?

**Comment by sysradium** [comment]
I have tried addressing this in my PR. Would be nice if anyone could have a look at it.
